### PR TITLE
Preserve user provided fonts in `fonts` when updating game version.

### LINKED ISSUE #424: Preserve `fonts` folder when updating game
**Is your feature request related to a problem? Please describe.**

I have fonts I use specifically for playing CDDA that are not installed on my system, but instead placed in the `fonts` folder in the root of the game directory. CDDA reads from this folder for additional, non-system fonts it can load via a user's `config/fonts.json` configuration file.

When updating the game via the launcher, the `fonts` directory is not preserved. To keep my fonts, I need to manually move font files from the `previous_version/fonts` directory into the new version's `fonts` directory.

**Describe the solution you'd like**

Preserve the `fonts` directory in a manner similar to how the `config` directory is preserved between versions.

**Describe alternatives you've considered**

* Managing this manually. I currently do, and it's a bit of a pain.
* Installing fonts. Not ideal, as I don't really need fonts I use for CDDA elsewhere. Many fonts used for CDDA are also not useful when fonts are scaled up in size past 12 or 14px (whiterbt, terminus, etc)

**Additional context**

I'm planning on making a PR to address this as I foray into contributing to this project. Let me know if there's anything I should keep in mind as I draft more PRs for review. :)

### PR DESCRIPTION
**Pull request by damien**
* Moved font preservation logic into UpdateGroupBox#preserve_custom_font
* Refactored UpdateGroupBox#preserve_custom_font to support additional font directories
* Added `font` to list of supported font directories, was previously only `data/font`

Fixes #424

## Note

I'm new to working on this repo. A few things I'm having trouble with:

1. There are no unit tests around this repository. I have no idea if these changes break the build, but I was able to verify the logic was sound in the python REPL.
2. I have no idea what version of python this repository uses. I'm using Python 3.8, but tried to stick to stuff that's compatible with Python 3.x
3. I'm currently unable to build this project, as not all of the app dependencies are fetchable via PyPi. There were some notes in the [BUILDING ](https://github.com/remyroy/CDDA-Game-Launcher/blob/master/BUILDING.md) document, but they sadly weren't enough for me to figure out how to get Python to load in the win32 binary deps the doc says we need to download manually.

## TLDR;

Merge with caution. If anybody can build this project, I'd appreciate any help verifying my changes are good. 🚀

### CODE REVIEW

**Comment by damien** [comment]
Caught a couple silly typos that I've fixed in updates to this PR. 👌

**Comment by remyroy** [comment]
I've updated the guide for running and building the launcher. I've also fixed an issue with Python >= 3.7 not working with a package that was broken. Can you test your PR now?

**Comment by damien** [comment]
I'll give it a shot later today, thanks for updating the docs! I'll let you know if I'm still having trouble; I'm happy to test this out as long as I know how. 😄

**Comment by remyroy** [comment]
If you need help with installing packages or running the launcher, let me know.

**Comment by damien** [comment]
Hey @remyroy!

First off, thanks for updating the docs _and_ the `requirements.txt` file! I was able to get to the point where I could compile and run the launcher locally using Visual Studio.

Back on topic, my suspicious were proven correct when I ran this branch and found that my code broke in a few scenarios. These issues have been fixed, so I'd love for you to check this out and see if this meets your contribution standards or point out any problems you find or see with this PR.

Cheers! 🍻

File: cddagl/ui/views/main.py [hunk 1]
```diff
@@ -2347,6 +2348,80 @@ def copy_next_soundpack(self):
...
+        logger.info('Mapped current to previous font dirs:')
+        logger.info('{}'.format(font_paths))
+
+        if (any(font_paths) and self.in_post_extraction):
+            logger.debug('Restoring custom fonts')
+            status_bar.showMessage(_('Restoring custom fonts'))
+
+            for font_dir, prev_font_dir in font_paths:
+                # Skip the dir if we have nothing to restore
+                if not prev_font_dir.is_dir():
```
Comment (remyroy) [inline cddagl/ui/views/main.py]: Is this needed if you already filter by `os.path.isdir(font_dir_map[1])` previously when defining font_paths?

File: cddagl/ui/views/main.py [hunk 2]
```diff
@@ -2347,6 +2348,80 @@ def copy_next_soundpack(self):
...
+                    pass
+
+                # Create a new font directory if it doesn't already exist
+                if not font_dir.is_dir():
+                    font_dir.mkdir(exist_ok=True)
+
+                previous_set = set(prev_font_dir.iterdir())
+                current_set  = set(font_dir.iterdir())
```
Comment (remyroy) [inline cddagl/ui/views/main.py]: It would be more appropriate to use `os.scandir()` for listing directories if you are going to check for various resulting entry attributes like testing if it's a file or a directory afterwards. The scandir() function returns directory entries along with file attribute information, giving better performance for many common use cases.

File: cddagl/ui/views/main.py (hunk 1 above)
Comment (damien) [reply cddagl/ui/views/main.py]: Fixed this by removing the `filter` clause in the `font_paths` variable assignment. I think doing this explicitly here is a bit easier to follow. 👌

File: cddagl/ui/views/main.py (hunk 2 above)
Comment (damien) [reply cddagl/ui/views/main.py]: Nice catch! Looking into this now! 🔥

Comment (damien) [reply cddagl/ui/views/main.py]: Fixed in a58e53ed5b1a6baf93d7af9679011eb0b99129de , let me know what you think @remyroy!

**Comment by damien** [comment]
Ready for a second round of review 🎉

File: cddagl/ui/views/main.py [hunk 3]
```diff
@@ -2347,6 +2348,81 @@ def copy_next_soundpack(self):
...
+
+                for entry in delta:
+                    source = prev_font_dir.joinpath(entry)
+                    target =      font_dir.joinpath(entry)
+
+                    logger.debug('  Restoring {}'.format(entry))
+                    logger.debug('    Source: {}'.format(source))
+                    logger.debug('    Target: {}'.format(target))
+
+                    if os.path.isfile(source):
```
Comment (remyroy) [inline cddagl/ui/views/main.py:2417]: This should probably be `if entry.is_file():` to reuse attributes obtained from `os.scandir()`.

File: cddagl/ui/views/main.py [hunk 4]
```diff
@@ -2347,6 +2348,81 @@ def copy_next_soundpack(self):
...
+                        logger.debug('    Copying file')
+                        shutil.copy2(source, target)
+                    elif os.path.isdir(source):
```
Comment (remyroy) [inline cddagl/ui/views/main.py:2420]: This should probably be `if entry.is_dir():` to reuse attributes obtained from `os.scandir()`.

File: cddagl/ui/views/main.py (hunk 3 above)
Comment (damien) [reply cddagl/ui/views/main.py:2417]: Good catch! I was testing for errors, but that wouldn't have caught me missing optimization improvements from reusing scandir metadata. I'll make that change sometime today. 👌

**Comment by remyroy** [comment]
I'll be making a new release for the launcher in the next few hours to include the recent 0.E release. I think the last 2 comments and related changes should be good for this PR. Let's try to get your PR in for the next release of the launcher.

**Comment by damien** [comment]
@remyroy Feedback implemented. I've changed instances where we were previously using `pathlib.Path` to instead preserve and leverage `os.DirEntry`'s cached filesystem metadata. Worked like a charm. 👌 

Trimmed out some debug logging that has outlived it's usefulness while I was in there. I think we're quite close to a final implementation of the font restoration logic. :)

**Comment by remyroy** [comment]
Seems good. Merging.
